### DISCUSSION TITLE
Workflow complete

### DISCUSSION THREAD

**Discussion by shubhamgoyal1402** [uber/cadence#5925] [Q&A]
How to complete a workflow in between if we have its workflow id .

**Comment by demirkayaender** [comment]
Do you mean terminate/cancel? Check following commands:

`cadence workflow terminate --help` and
`cadence workflow cancel --help`

**Comment by shubhamgoyal1402** [comment]
No don't terminate I want to complete the workflow, I have the context, run
id and workflow id how to achieve by using this.

**Comment by demirkayaender** [comment]
I think I'm still missing something :) 
You cannot complete workflows externally. It only completes based on the workflow logic. 
Do you maybe mean you want to signal the workflow?

You can add a [signal](https://cadenceworkflow.io/docs/java-client/signals/#implement-signal-handler-in-workflow) for your workflow and complete your workflow upon receiving the signal in your workflow logic.

You can send a signal to a workflow when you know the domain name, workflow id and the signal name.

**Comment by shubhamgoyal1402** [comment]
Yes exactly I want to achieve this, but can you guide how to send signal
and how to receive that signal in an activity, and the signal I'm sending
is out of the workflow like a function working in background not included
in workflow

**Reply by demirkayaender** [reply]
You cannot receive signals inside an activity, they can only be received in the workflow code (outside the activity). Feel free to check our samples ([example](https://github.com/uber-common/cadence-samples/blob/master/new_samples/workflows/signal_greeter_multi_language.go)).

You normally don't want to wait inside an activity. Your activities should always be processing something and the "waiting" should be inside the workflow code via signals and timers. 

If you have to use an activity, then you need to implement receiving signals by yourself inside the activity as Cadence doesn't have access to your activity. If you can do that, then your activity can immediately finish and return something indicating workflow needs to complete. Then in your workflow code, based on that return value, you can complete your workflow.

Does that answer your question?

**Comment by shubhamgoyal1402** [comment]
how to send signal though

**Reply by demirkayaender** [reply]
You can either send it from the CLI or programmatically.
* For CLI, check `cadence workflow signal --help`
* For program, check [this test](https://github.com/uber-common/cadence-samples/blob/master/cmd/samples/recipes/signalcounter/workflow_test.go#L36) as an example.

**Comment by shubhamgoyal1402** [comment]
can you write me a code where a function is sending signal and the worflow
is revcing it

**Reply by demirkayaender** [reply]
Here's a sample about signals:
* workflow that receives signals: https://github.com/uber-common/cadence-samples/blob/master/cmd/samples/recipes/signalcounter/signal_counter_workflow.go#L31
* test that's sending signal https://github.com/uber-common/cadence-samples/blob/master/cmd/samples/recipes/signalcounter/workflow_test.go#L36

**Comment by shubhamgoyal1402** [comment]
any other way to send signal to a workflow without using cmd line

**Reply by demirkayaender** [reply]
The second link in the response above showing how to send signals programmatically on line 36.

This is the signal interface where you can see the API doc: https://github.com/uber-go/cadence-client/blob/master/client/client.go#L168

**Comment by shubhamgoyal1402** [comment]
Actually i have a function in which it dequeues from a queue, this is the
function in main.go file. when we dequeue we can get workflow id , run id,
ctx   of a workflow ,  now i want to send signal to a workflow which is in
customer.go file waiting to receive a signal. how can i achieve this?

**Reply by demirkayaender** [reply]
The links above exactly do that. You need to call the SignalWorkflow API from the Cadence Client object. 
Here are all the samples about signaling: https://github.com/search?q=repo%3Auber-common%2Fcadence-samples+signalworkflow&type=code

Here's an example StackOverflow entry too: https://stackoverflow.com/questions/56568871/sending-external-events-to-a-workflow (search SignalWorkflow in the page)

You can find other examples by searching "cadence signalworkflow" too. There are several blogs talking about signals as well.

**Comment by shubhamgoyal1402** [comment]
I'll check them out, but sir i am really stuck on this , I will
implement this once and get back at you


Thank you

**Reply by demirkayaender** [reply]
Especially the last links I provided should really help. Not sure what else I can provide you at this point :) 
I guess you can alternatively try ChatGPT code-gen too: https://chat.openai.com/share/ebc37a80-2523-4898-9c10-8b89f92251d1

**Reply by demirkayaender** [reply]
This was the GPT's output in case its retention expires:

Here's a sample Cadence workflow in Go that accepts a signal to complete and a code to signal that workflow.


    package main

    import (
      "context"
      "fmt"
      "go.uber.org/cadence/activity"
      "go.uber.org/cadence/workflow"
      "go.uber.org/zap"
      "go.uber.org/cadence/worker"
    )

    // WorkflowName is the name of the workflow
    const WorkflowName = "SampleWorkflow"

    // SignalName is the name of the signal
    const SignalName = "CompleteSignal"

    // SampleWorkflow is a sample Cadence workflow
    func SampleWorkflow(ctx workflow.Context) error {
      logger := workflow.GetLogger(ctx)
      logger.Info("SampleWorkflow started")

      var completed bool
      // This channel will be used to receive the signal
      signalChannel := workflow.GetSignalChannel(ctx, SignalName)
      // Wait for the signal
      workflow.Go(ctx, func(ctx workflow.Context) {
        var signalData string
        // This will block until the signal is received
        signalChannel.Receive(ctx, &signalData)
        logger.Info("Received signal", zap.String("SignalData", signalData))
        completed = true
      })

      // This loop will keep waiting for the signal
      for !completed {
        // Sleep for a short duration before checking again
        workflow.Sleep(ctx, 1)
      }

      logger.Info("Workflow completed")
      return nil
    }

    // SignalWorkflow is used to signal the workflow
    func SignalWorkflow(ctx context.Context, workflowID, signalData string) error {
      // Create a Cadence client
      client, err := workflow.NewClient(ctx)
      if err != nil {
        return err
      }
      defer client.Close()

      // Use the client to signal the workflow
      signalClient := client.SignalWithStartWorkflow(ctx, workflowID)
      err = signalClient.SignalWorkflow(SignalName, signalData, workflowID, SampleWorkflow, nil)
      if err != nil {
        return err
      }
      return nil
    }

    // RegisterWorker registers the workflow and activity with the worker
    func RegisterWorker(w worker.Worker) {
      w.RegisterWorkflow(SampleWorkflow)
    }

    func main() {
      // Create a Worker.
      workerOptions := worker.Options{
        Logger:                   zap.NewNop(),
        MaxConcurrentActivityTaskPollers: 5,
      }

      worker := worker.New(nil, "sample-task-list", workerOptions)
      RegisterWorker(worker)

      // Start the Worker.
      err := worker.Start()
      if err != nil {
        panic(fmt.Sprintf("unable to start worker: %v", err))
      }
    }


In this sample, SampleWorkflow waits for a signal named CompleteSignal. You can use the SignalWorkflow function to signal this workflow.

Make sure to include the necessary dependencies in your go.mod:



    module cadence_sample

    go 1.17

    require (
      github.com/uber/cadence-client v0.2.3
      go.uber.org/cadence v0.8.0
      go.uber.org/zap v1.17.0
    )


And install them using:
`go mod tidy`

**Comment by shubhamgoyal1402** [comment]
c, err := client.NewClient(client.Options{})
        if err != nil {
            log.Fatalln("unable to create Temporal client", err)
        }

The arguments for this is not complete it is saying

[image: image.png]

**Reply by demirkayaender** [reply]
here's an example way to build client: https://github.com/uber-common/cadence-samples/blob/df6f7bdba978d6565ad78e9f86d9cd31dfac9f78/new_samples/worker/worker.go#L72

You may need to ready/check the samples I sent above.

**Comment by shubhamgoyal1402** [comment]
Not able to figure out, if it's possible to get on a meet? It'll be really
appreciated as I am working hard on this project

**Comment by shubhamgoyal1402** [comment]
Sir thank you for your time and co-operation. Was able to figure it out
with your links and help.

Thank you

**Reply by demirkayaender** [reply]
I'm happy to hear you were able to figure it out. We are also adding an explicit sample for your case. You'll see it soon in our samples.
